fix: queue CCIP transactions by dest chain, not by lane

roman-kashitsyn opened the pull request:
## Motivation

Our current queuing strategy is based on the externalJobID today: We prune the unstarted queue on a per-job basis, which in CCIP is per lane. So the queue of unstarted transactions per chain can hit the maximum size (250), and then we will stop enqueuing new transactions.

## Solution

Queue transactions by destination chain instead. This design ensures that we are less likely to hit the per-chain queue limit under normal operation: CCIP commit and exec jobs will use at most [`DefaultTransactionQueueDepth`](https://github.com/smartcontractkit/chainlink/blob/develop/docs/CONFIG.md#defaulttransactionqueuedepth-2) (currently [150](https://github.com/smartcontractkit/infra-k8s/blob/2ce8d0e33303528c7ed41c314a7adb0e021070fa/projects/chainlink/files/chainlink-clusters/clc-ocr-multichain-ccip-omega/config.yaml#L66)) slots per chain. The total limit of unconfirmed transactions per key is [MaxQueued](https://github.com/smartcontractkit/chainlink/blob/develop/docs/CONFIG.md#maxqueued) (currently the default 250).

roman-kashitsyn committed:
fix: queue CCIP transactions by dest chain, not by lane

roman-kashitsyn committed:
Merge remote-tracking branch 'origin/ccip-develop' into CCIP-1505-custom-subject

roman-kashitsyn committed:
undo an unnecessary core change

roman-kashitsyn committed:
remove more unrelated changes

roman-kashitsyn committed:
add a test for subject computation

AnieeG commented on core/services/relay/evm/ccip_test.go:
```diff
@@ -0,0 +1,13 @@
+)
+
+func Test_CCIPSubjectUUID(t *testing.T) {
+	assert.Equal(t, chainToUUID(big.NewInt(1)), chainToUUID(big.NewInt(1)))
```
is there any way to assert this against already calculated expectation? If there is any bug in `chainToUUID` it will go unnoticed with this approach of assertion.

roman-kashitsyn replied:
Good point, I’ve added a test case with a hard-coded value. This check ensures that the function produces stable outputs across the node runs.

roman-kashitsyn committed:
add a test case

roman-kashitsyn committed:
split queue subjects per plugin

roman-kashitsyn committed:
add a changeset (?)

connorwstein commented on .changeset/six-wolves-relax.md:
```diff
@@ -0,0 +1,5 @@
+---
+"ccip": patch
```
?

roman-kashitsyn replied:
The CI made me do it: https://github.com/smartcontractkit/ccip/pull/619#issuecomment-2003995763

mateusz-sekara reviewed (commented):
> Queue transactions by destination chain instead. This design ensures that we never hit the per-chain queue limit under normal operation: CCIP commit and exec jobs will use at most 150 slots each, below the total limit of 500

@roman-kashitsyn @connorwstein is it sth that might become a problem during chain expansion? I mean, as we scale, do we need to bump this number?

makramkd commented:
> CCIP commit and exec jobs will use at most 150 slots each, below the total limit of 500

I believe queue size is configurable (https://github.com/smartcontractkit/chainlink/blob/develop/docs/CONFIG.md#maxqueued) or are these numbers configured elsewhere / hardcoded?

roman-kashitsyn commented:
> are these numbers configured elsewhere / hardcoded

These numbers are configured in the TOML config.

The [`DefaultTransactionQueueDepth`](https://github.com/smartcontractkit/chainlink/blob/develop/docs/CONFIG.md#defaulttransactionqueuedepth-2) option sets the per-plugin limit, which is [currently 150](https://github.com/smartcontractkit/infra-k8s/blob/2ce8d0e33303528c7ed41c314a7adb0e021070fa/projects/chainlink/files/chainlink-clusters/clc-ocr-multichain-ccip-omega/config.yaml#L66).

The total limit is indeed 250 by default, not 500. I don‘t know where I got 500; that’s probably my memory failure. Great catch; I updated the description.

I also reverted to the original approach with a shared per-chain limit, so the total queue size is 150 < 250.

roman-kashitsyn committed:
revert to the first design with fixed seed
